### DISCUSSION TITLE
How to specify order of `dev` tasks?

### DISCUSSION THREAD

**Discussion by mayank99** [vercel/turborepo#1299] [Help] [answered]
Consider this common setup where an app `B` depends on a package `A`:
```
packages/A/package.json
apps/B/package.json
```

The ideal scenario when running `turbo run dev` would be that it starts a dev server such that making a change inside package `A` should automatically trigger HMR in the app `B`.

But to start the dev server of `B`, it needs to have an initial set of built files available from package `A`.

So the order needs to be as follows:
1. Run `dev` on package `A`
2. Wait for the command to finish initial compilation and go into "watch mode".
3. Start `dev` on app `B`

I tried doing this in my pipeline but it doesn't work because the `dev` command doesn't exit:
```json
"dev": {
  "dependsOn": ["^dev"]
}
```

**Comment by ThibautMarechal** [comment, marked as answer]
Since the ``dev`` pipeline does not exit, turbo cannot wait for something that isn't over yet.

You could do something like that in your package B's package.json :

```json
{
  "name": "B",
  "scripts": {
    "dev": "run-s wait-dev remix dev",
    "wait-dev": "wait-on ../A/lib/index.js"
  }
}
```
using the packages  [wait-on](https://www.npmjs.com/package/wait-on) and [npm-run-all](https://www.npmjs.com/package/npm-run-all)

It's working, but it feels too hacky to me 😄

**Reply by mayank99** [reply]
Thank you for the suggestion. That sounds reliable enough (despite feeling hacky) and I doubt turbo could offer anything better, so I will mark this as answer.

**Comment by thenbe** [comment]
This seems to work. Turbo will wait for the `backend` to become healthy before starting the `site`. This is done by creating a [specific workspace-task](https://turborepo.org/docs/core-concepts/running-tasks#specific-workspace-tasks).

### turbo.json
```json
  "pipeline": {
    "health-check": {
      "cache": false
    },
    "site#dev": {
      "cache": false,
      "dependsOn": [
        "backend#health-check"
      ]
    }
  }
```

### packages/backend/package.json
```json
"scripts": {
  "health-check": "./health-probe.sh http://localhost:8080 10"
}
```

### health-probe.sh
```bash
#!/bin/bash

# The script takes two arguments, the url to check and the timeout. 
# It then checks the health endpoint every second until the timeout is reached. 
# If the health endpoint returns a 200 status code, the script exits with a 0 status code.
# If the timeout is reached, the script exits with a 1 status code.

# Usage: health-probe.sh <url> <timeout>
# Example: health-probe.sh http://localhost:8080/health 60

url=$1
timeout=$2
interval=1
elapsed=0

while [ $elapsed -lt $timeout ]; do
  status=$(curl -s -o /dev/null -w "%{http_code}" $url)
  if [ $status -eq 200 ]; then
    echo "API is healthy"
    exit 0
  fi
  echo "API not ready, waiting $interval seconds"
  sleep $interval
  elapsed=$((elapsed + interval))
done

echo "API not ready after $timeout seconds"
exit 1
```

**Comment by anthonyshew** [comment]
I see "watch mode" mentioned here. We've shipped this in 2.0: https://turbo.build/repo/docs/reference/watch

Feel free to mark as answer and close if this fits what you're looking for.
